api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/tidwall/sjson"
+	"net/http"
+	"strings"
+
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+const ReferenceInstanceFilterName = "ReferenceInstanceFilter"
+
+// serviceInstanceVisibilityFilter ensures that the tenant making the provisioning/update request
+// has the necessary visibilities - i.e. that he has the right to consume the requested plan.
+type referenceInstanceFilter struct {
+	repository       storage.TransactionalRepository
+	tenantIdentifier string
+}
+
+func NewReferenceInstanceFilter(repository storage.TransactionalRepository, tenantIdentifier string) *referenceInstanceFilter {
+	return &referenceInstanceFilter{
+		repository:       repository,
+		tenantIdentifier: tenantIdentifier,
+	}
+}
+
+func (*referenceInstanceFilter) Name() string {
+	return ReferenceInstanceFilterName
+}
+
+func (rif *referenceInstanceFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	switch req.Request.Method {
+	case http.MethodGet:
+		if strings.Contains(req.RequestURI, "/parameters") {
+			return rif.handleGetParameters(req, next)
+		}
+	case http.MethodPost:
+		return rif.handleProvision(req, next)
+	case http.MethodPatch:
+		return rif.handleServiceUpdate(req, next)
+	}
+	return next.Handle(req)
+}
+
+func (rif *referenceInstanceFilter) handleGetParameters(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	instanceID := req.PathParams[web.PathParamResourceID]
+
+	dbInstanceObject, err := storage.GetObjectByField(ctx, rif.repository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	instance := dbInstanceObject.(*types.ServiceInstance)
+	if instance.ReferencedInstanceID == "" {
+		return next.Handle(req)
+	}
+
+	var marshal []byte
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json")
+	body := map[string]string{
+		instance_sharing.ReferencedInstanceIDKey: instance.ReferencedInstanceID,
+	}
+	marshal, err = json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	return &web.Response{
+		Body:       marshal,
+		StatusCode: http.StatusOK,
+		Header:     headers,
+	}, nil
+}
+
+func (*referenceInstanceFilter) FilterMatchers() []web.FilterMatcher {
+	return []web.FilterMatcher{
+		{
+			Matchers: []web.Matcher{
+				web.Path(web.ServiceInstancesURL + "/**"),
+				web.Methods(http.MethodPost, http.MethodPatch, http.MethodGet),
+			},
+		},
+	}
+}
+
+func (rif *referenceInstanceFilter) handleProvision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := storage.IsReferencePlan(ctx, rif.repository, types.ServicePlanType.String(), "id", servicePlanID)
+	// If plan not found on provisioning, allow sm to handle the issue
+	if err == util.ErrNotFoundInStorage {
+		return next.Handle(req)
+	}
+	if err != nil {
+		return nil, err // handled by the IsReference function
```

Comment:
I am not sure it is handled in the IsReferencePlan function like the comment says, don't you need a handleStorageError in there?

Reply:
We've fixed the IsReferencePlan so it returns the handleStorageError, waiting for tests to run.